2024/articles/omp-santa.pod

```diff
@@ -0,0 +1,339 @@
+  -39.1505435434735 -14.0167682855066 25391.598017652
+  -14.701859640773 33.3797684668173 28958.4392020613
+  22.7094543766397 28.9295184727116 29847.2350918171
+  58.9987788505186 -87.6847921052664 29544.1317147911
+  83.6874173858257 -149.058764882263 22417.9224396509
+  -8.01336267852399 97.8876777856595 27879.4674084787
+  -55.3867650906297 -107.353651427755 21389.9702111095
+  -78.8588152992001 -155.558248147836 25430.0402744441
+  9.88995820014878 -0.204367766261981 20832.5618074863
+  -76.8565255868645 -14.4804333171123 27013.5287117141
+  -0.16890065869049 -40.7974093702016 22440.2960018416
+  8.56759320194605 14.0242190926548 24229.1350707098
+  89.3116725410715 19.3710347706399 28181.9446348641
+  ...
+
+=end data
+
+The distance formula was computationally expensive, involving square roots and
+trigonometry. But by using OpenMP, the task could be split into multiple OS
+threads, each calculating the distance to one satellite at the same time. The
+results would then be aggregated, and the sleigh's GPS would know exactly
+where to direct its heading.
+
+=begin c 
+
+  // Function to calculate Euclidean distance between two points (x1, y1, z1) and (x2, y2, z2)
+  double calculate_distance(double x1, double y1, double z1, double x2, double y2, double z2) {
+      return sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2) + pow(z2 - z1, 2));
+  }
+
+=end c
+
+The code would parallize the work by dividing the list of satellites
+among a number of I<pthreads> using OpenMP's C<#pragma omp for>
+construct! C<OpenMP::Simple> handles the required C<< #include <omp.h> >>
+and makes it easy to query the environment for information, such as C<OMP_NUM_THREADS>.
+
+=begin c 
+
+  PerlOMP_GETENV_BASIC // read common environmental variables, provided by OpenMP::Simple 
+
+  // Start parallel region
+  #pragma omp parallel shared(_distances)
+  {
+    // Parallel for loop to compute distances in parallel
+    #pragma omp for
+    for (int i = 0; i < satellite_count; i++) {
+        _distances[i] = 0.0;  // Initialize
+
+        double sat_x, sat_y, sat_z;
+
+        // Convert satellite's geographic coordinates to Cartesian coordinates
+        geo_to_cartesian(sat_latitudes[i], sat_longitudes[i], sat_altitudes[i], &sat_x, &sat_y, &sat_z);
+
+        // Calculate the distance from the sleigh to this satellite
+        _distances[i] = calculate_distance(sleigh_x, sleigh_y, sleigh_z, sat_x, sat_y, sat_z);
+    }
+  }
+
+=end c
+
+Jingles quickly typed up a solution in Perl, his favorite programming
+language. The solution was simple but critically offloaded the triangulation
+computations to C<Inline::C> code containing OpenMP directives, using the
+C<OpenMP> module on I<CPAN>.
+
+=begin perl
+
+  use strict;
+  use warnings;
+  
+  use OpenMP;
+     
+  use Inline (
+      C    => 'DATA',
+      with => qw/OpenMP::Simple/,
+  );
+     
+  my $omp = OpenMP->new;
+  
+  $omp->env->omp_num_threads($ARGV[0] // 8); # accept number of threads via commandline
```

Comment:
Maybe read from some environment variable the actual number of threads in the machine?